providers/base/bin/camera_test.py

```diff
@@ -282,89 +370,204 @@ def _setup(self, sink=None):
         """
-        Captures an image to a given filename.  width and height specify the
-        image size and quiet controls whether the image is displayed to the
-        user (quiet = True means do not display image).
+
+        # Try to take a picture with fswebcam
+        use_fswebcam = True
+        use_gstreamer = False
+
+        if use_fswebcam:
+            result = self._capture_image_fswebcam(
+                filename, width, height, pixelformat
+            )
+            if not result:
+                print("Failed to capture image with fswebcam, using gstreamer")
+                use_gstreamer = True
+
+        # If fswebcam fails, try with gstreamer
+        if use_gstreamer:
+            self._capture_image_gstreamer(filename, width, height, pixelformat)
+            print("Image saved to %s" % filename)
+        if not self.quiet:
+            self._display_image(filename, width, height)
+
+    def _capture_image_fswebcam(self, filename, width, height, pixelformat):
+        """
+        Simple wrapper around fswebcam to capture an image
         """
         command = [
             "fswebcam",
             "-D 1",
             "-S 50",
             "--no-banner",
             "-d",
-            self.args.device,
+            self.device,
             "-r",
             "%dx%d" % (width, height),
             filename,
         ]
-        use_camerabin = False
         if pixelformat:
             if "MJPG" == pixelformat:  # special tweak for fswebcam
                 pixelformat = "MJPEG"
             command.extend(["-p", pixelformat])
-
         try:
             check_call(command, stdout=open(os.devnull, "w"), stderr=STDOUT)
             if os.path.getsize(filename) == 0:
-                use_camerabin = True
+                return False
         except (CalledProcessError, OSError):
-            use_camerabin = True
-        if use_camerabin:
-            self._setup(sink="fakesink")
-            GLib.timeout_add_seconds(3, self._take_photo, filename)
-            GLib.timeout_add_seconds(4, self._stop)
-            Gtk.main()
-        if not quiet:
-            stage = Clutter.Stage()
-            stage.set_title("Camera still picture test")
-            stage.set_size(width, height)
-            stage.connect("destroy", self._on_destroy)
-            Clutter.threads_add_timeout(0, 10000, self._on_destroy, None, None)
-            still_texture = Clutter.Texture.new_from_file(filename)
-            stage.add_actor(still_texture)
-            stage.show()
-            Clutter.main()
+            return False
+        return True
+
+    def _capture_image_gstreamer(self, filename, width, height, pixelformat):
+        """
+        Setup the gstreamer pipeline to capture an image.
+        This pipeline consists of the following elements:
+        - v4l2src: Capture video from a V4L2 device
+        - caps: Filter the video stream to the desired resolution
```

Comment:
So caps is (my best understanding here) a sort of "native" or "raw" capture to a set resolution. Can we rename it `raw_capture` so that it is clear what it is doing? (also makes the `rgb_capture` name clearer in my suggestion below)

Reply:
So the capture is performed by the `v4l2src` element. The `caps` is an element that contains the information to handle a specific media type.
It is a bit confusing, but AFAIK, it is how Gstreamer handles different media types. You have to define the different caps values and gstreamer will try to set the video capture and sink accordingly.
It does not capture anything by itself but tells the `v4l2src` element how to capture it in the right format.

Reply:
I have since encountered this naming format in another unrelated project, lets keep it as is